File: codigo/python/classical.py
"""
Implementação dos Algoritmos Clássicos de Ordenação da Literatura.
Cada função retorna (lista_ordenada, comparacoes, movimentacoes).
"""

from typing import Any, List, Tuple


def bubble_sort(arr: List[Any]) -> Tuple[List[Any], int, int]:
    """
    Bubble Sort com otimização de parada antecipada.
    Complexidade: Melhor O(n), Pior O(n²), Médio O(n²).
    Estabilidade: Estável.
    """
    a = list(arr)
    n = len(a)
    comps = 0
    moves = 0

    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            comps += 1
            if a[j] > a[j + 1]:
                a[j], a[j + 1] = a[j + 1], a[j]
                moves += 2
                swapped = True
        if not swapped:
            break

    return a, comps, moves


def selection_sort(arr: List[Any]) -> Tuple[List[Any], int, int]:
    """
    Selection Sort.
    Complexidade: Melhor O(n²), Pior O(n²), Médio O(n²).
    Estabilidade: Não estável (versão in-place com trocas).
    """
    a = list(arr)
    n = len(a)
    comps = 0
    moves = 0

    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            comps += 1
            if a[j] < a[min_idx]:
                min_idx = j
        if min_idx != i:
            a[i], a[min_idx] = a[min_idx], a[i]
            moves += 2

    return a, comps, moves


def insertion_sort(arr: List[Any]) -> Tuple[List[Any], int, int]:
    """
    Insertion Sort linear.
    Complexidade: Melhor O(n), Pior O(n²), Médio O(n²).
    Estabilidade: Estável.
    """
    a = list(arr)
    n = len(a)
    comps = 0
    moves = 0

    for i in range(1, n):
        key = a[i]
        moves += 1
        j = i - 1
        while j >= 0:
            comps += 1
            if a[j] > key:
                a[j + 1] = a[j]
                moves += 1
                j -= 1
            else:
                break
        a[j + 1] = key
        moves += 1

    return a, comps, moves


def merge_sort(arr: List[Any]) -> Tuple[List[Any], int, int]:
    """
    Merge Sort clássico (Divisão e Conquista).
    Complexidade: Melhor O(n log n), Pior O(n log n), Médio O(n log n).
    Espaço auxiliar: O(n).
    Estabilidade: Estável.
    """
    a = list(arr)
    n = len(a)
    if n <= 1:
        return a, 0, 0

    comps = [0]
    moves = [0]

    def _merge_sort_rec(lst: List[Any]) -> List[Any]:
        if len(lst) <= 1:
            return lst

        mid = len(lst) // 2
        left = _merge_sort_rec(lst[:mid])
        right = _merge_sort_rec(lst[mid:])

        merged = []
        i = j = 0
        while i < len(left) and j < len(right):
            comps[0] += 1
            if left[i] <= right[j]:
                merged.append(left[i])
                moves[0] += 1
                i += 1
            else:
                merged.append(right[j])
                moves[0] += 1
                j += 1

        while i < len(left):
            merged.append(left[i])
            moves[0] += 1
            i += 1

        while j < len(right):
            merged.append(right[j])
            moves[0] += 1
            j += 1

        return merged

    result = _merge_sort_rec(a)
    return result, comps[0], moves[0]


def quick_sort(arr: List[Any]) -> Tuple[List[Any], int, int]:
    """
    Quick Sort com partição de Hoare e escolha de pivô mediana de três.
    Complexidade: Melhor O(n log n), Pior O(n²), Médio O(n log n).
    Estabilidade: Não estável.
    """
    a = list(arr)
    n = len(a)
    if n <= 1:
        return a, 0, 0

    comps = [0]
    moves = [0]

    def _median_of_three(low: int, high: int) -> int:
        mid = (low + high) // 2
        comps[0] += 3
        if a[low] > a[mid]:
            a[low], a[mid] = a[mid], a[low]
            moves[0] += 2
        if a[low] > a[high]:
            a[low], a[high] = a[high], a[low]
            moves[0] += 2
        if a[mid] > a[high]:
            a[mid], a[high] = a[high], a[mid]
            moves[0] += 2
        return mid

    def _quick_sort_rec(low: int, high: int) -> None:
        if low >= high:
            return

        # Mediana de 3 para seleção de pivô robusta
        pivot_idx = _median_of_three(low, high)
        pivot = a[pivot_idx]

        i = low
        j = high

        while i <= j:
            while True:
                comps[0] += 1
                if a[i] < pivot:
                    i += 1
                else:
                    break
            while True:
                comps[0] += 1
                if a[j] > pivot:
                    j -= 1
                else:
                    break
            if i <= j:
                if i != j:
                    a[i], a[j] = a[j], a[i]
                    moves[0] += 2
                i += 1
                j -= 1

        if low < j:
            _quick_sort_rec(low, j)
        if i < high:
            _quick_sort_rec(i, high)

    _quick_sort_rec(0, n - 1)
    return a, comps[0], moves[0]

File: codigo/cpp/classical.cpp
#include "classical.hpp"
#include <algorithm>

SortResult bubble_sort(std::vector<int> arr) {
    SortResult res;
    res.data = std::move(arr);
    res.comparisons = 0;
    res.moves = 0;

    size_t n = res.data.size();
    if (n <= 1) return res;

    for (size_t i = 0; i < n; ++i) {
        bool swapped = false;
        for (size_t j = 0; j < n - i - 1; ++j) {
            res.comparisons++;
            if (res.data[j] > res.data[j + 1]) {
                std::swap(res.data[j], res.data[j + 1]);
                res.moves += 2;
                swapped = true;
            }
        }
        if (!swapped) break;
    }
    return res;
}

SortResult selection_sort(std::vector<int> arr) {
    SortResult res;
    res.data = std::move(arr);
    res.comparisons = 0;
    res.moves = 0;

    size_t n = res.data.size();
    if (n <= 1) return res;

    for (size_t i = 0; i < n; ++i) {
        size_t min_idx = i;
        for (size_t j = i + 1; j < n; ++j) {
            res.comparisons++;
            if (res.data[j] < res.data[min_idx]) {
                min_idx = j;
            }
        }
        if (min_idx != i) {
            std::swap(res.data[i], res.data[min_idx]);
            res.moves += 2;
        }
    }
    return res;
}

SortResult insertion_sort(std::vector<int> arr) {
    SortResult res;
    res.data = std::move(arr);
    res.comparisons = 0;
    res.moves = 0;

    size_t n = res.data.size();
    if (n <= 1) return res;

    for (size_t i = 1; i < n; ++i) {
        int key = res.data[i];
        res.moves++;
        int64_t j = static_cast<int64_t>(i) - 1;
        while (j >= 0) {
            res.comparisons++;
            if (res.data[j] > key) {
                res.data[j + 1] = res.data[j];
                res.moves++;
                j--;
            } else {
                break;
            }
        }
        res.data[j + 1] = key;
        res.moves++;
    }
    return res;
}

static void merge_sort_rec(std::vector<int>& arr, std::vector<int>& temp,
                           size_t left, size_t right,
                           uint64_t& comps, uint64_t& moves) {
    if (left >= right) return;

    size_t mid = left + (right - left) / 2;
    merge_sort_rec(arr, temp, left, mid, comps, moves);
    merge_sort_rec(arr, temp, mid + 1, right, comps, moves);

    size_t i = left, j = mid + 1, k = left;
    while (i <= mid && j <= right) {
        comps++;
        if (arr[i] <= arr[j]) {
            temp[k++] = arr[i++];
            moves++;
        } else {
            temp[k++] = arr[j++];
            moves++;
        }
    }
    while (i <= mid) {
        temp[k++] = arr[i++];
        moves++;
    }
    while (j <= right) {
        temp[k++] = arr[j++];
        moves++;
    }
    for (size_t idx = left; idx <= right; ++idx) {
        arr[idx] = temp[idx];
        moves++;
    }
}

SortResult merge_sort(std::vector<int> arr) {
    SortResult res;
    res.data = std::move(arr);
    res.comparisons = 0;
    res.moves = 0;

    if (res.data.size() <= 1) return res;

    std::vector<int> temp(res.data.size());
    merge_sort_rec(res.data, temp, 0, res.data.size() - 1, res.comparisons, res.moves);
    return res;
}

static void quick_sort_rec(std::vector<int>& arr, int64_t low, int64_t high,
                           uint64_t& comps, uint64_t& moves) {
    if (low >= high) return;

    int64_t mid = low + (high - low) / 2;
    // Median-of-three
    comps += 3;
    if (arr[low] > arr[mid]) { std::swap(arr[low], arr[mid]); moves += 2; }
    if (arr[low] > arr[high]) { std::swap(arr[low], arr[high]); moves += 2; }
    if (arr[mid] > arr[high]) { std::swap(arr[mid], arr[high]); moves += 2; }

    int pivot = arr[mid];
    int64_t i = low;
    int64_t j = high;

    while (i <= j) {
        while (true) {
            comps++;
            if (arr[i] < pivot) i++; else break;
        }
        while (true) {
            comps++;
            if (arr[j] > pivot) j--; else break;
        }
        if (i <= j) {
            if (i != j) {
                std::swap(arr[i], arr[j]);
                moves += 2;
            }
            i++;
            j--;
        }
    }

    if (low < j) quick_sort_rec(arr, low, j, comps, moves);
    if (i < high) quick_sort_rec(arr, i, high, comps, moves);
}

SortResult quick_sort(std::vector<int> arr) {
    SortResult res;
    res.data = std::move(arr);
    res.comparisons = 0;
    res.moves = 0;

    if (res.data.size() <= 1) return res;
    quick_sort_rec(res.data, 0, static_cast<int64_t>(res.data.size()) - 1, res.comparisons, res.moves);
    return res;
}
